alacritty/src/renderer/box_drawing.rs

```diff
@@ -0,0 +1,584 @@
+    fn draw_rect(&mut self, x: f32, y: f32, width: f32, height: f32, color: RGBPixel) {
+        let start_x = x as usize;
+        let end_x = min((x + width) as usize, self.width);
+        let start_y = y as usize;
+        let end_y = min((y + height) as usize, self.height);
+        for y in start_y..end_y {
+            let y = y * self.width;
+            self.buffer[start_x + y..end_x + y].fill(color);
+        }
+    }
+
+    /// Naive arbitrary line drawing from (`from_x`, `from_y`) to (`to_x`, `to_y`).
+    fn draw_line(&mut self, from_x: f32, from_y: f32, to_x: f32, to_y: f32) {
+        let d_x = to_x - from_x;
+        let d_y = to_y - from_y;
+        for x in from_x as usize..=to_x as usize {
+            let y = from_y + d_y * (x as f32 - from_x) / d_x;
+            let y = y.clamp(0., self.height as f32 - 1.);
+            let index = min(x + y as usize * self.width, self.buffer.len() - 1);
+            self.buffer[index] = COLOR_FILL;
+        }
+    }
+
+    /// Draws an arc from `(0, self.center_v())` to `(self.center_h(), 0)`.
+    ///
+    /// You can mirror Arc in whichever direction you'd like.
+    fn draw_arc_centered(&mut self, stroke_size: usize) {
+        let v_line_bounds = self.v_line_bounds(self.center_h(), stroke_size);
+        let v_line_bounds = (v_line_bounds.0 as usize + 1, v_line_bounds.1 as usize);
+        let h_line_bounds = self.h_line_bounds(self.center_v(), stroke_size);
+        let h_line_bounds = (h_line_bounds.0 as usize + 1, h_line_bounds.1 as usize);
+
+        for (to_x, from_y) in (v_line_bounds.0..=v_line_bounds.1)
+            .into_iter()
+            .zip((h_line_bounds.0..=h_line_bounds.1).into_iter())
+        {
+            let d1 = to_x as f32;
+            let d2 = from_y as f32;
+
+            let mut y = from_y as f32;
+            while y >= 0. {
+                let x = f32::sqrt(d2 * d2 - y * y) * d1 / d2;
+                let x_r = min(x as usize, v_line_bounds.1 as usize - 1);
+                let y_r = min(y as usize, h_line_bounds.1 as usize - 1);
+                let index = min(x_r + y_r as usize * self.width, self.buffer.len() - 1);
+                self.buffer[index] = COLOR_FILL;
+                y -= 0.1;
+            }
+        }
+    }
+
+    /// Fills the `Canvas` with the given `Color`.
+    fn fill(&mut self, color: RGBPixel) {
+        self.buffer.fill(color);
+    }
+
+    /// Consumes `Canvas` and returns its underlying storage as raw byte vector.
+    fn into_raw_buffer(self) -> Vec<u8> {
+        // SAFETY This is safe since we use `repr(packed)` on `Pixel` struct for underlying storage
+        // of the `Canvas` buffer which consists of three u8 values.
+        unsafe {
+            let capacity = self.buffer.capacity();
+            let len = self.buffer.len() * 3;
+            let buf = self.buffer.as_ptr() as *mut u8;
+            std::mem::forget(self.buffer);
+            Vec::from_raw_parts(buf, len, capacity)
```

Comment:
I'm not sure the capacity specified here is correct? Because ultimately the capacity of the old vec is going to be a third of the capacity of the new one? Length **must** be equal or less than capacity and since you're multiplying the length of the original vector by 3, if the capacity is the same as the length this will not hold true.

Reply:
Oh, I somehow assumed that capacity returns bytes, but it's also in elements....